Downtime Monitoring: Create Bulk Edit Menu

DustinHartzler opened the pull request:
#### Proposed Changes

This PR implements a bulk edit menu for large and small screen views.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/jetpack-dashboard-downtime-monitoring-bulk-action-menu` and `yarn start-jetpack-cloud`
2. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
3. Verify that you can now see the `Edit All` button. Clicking the button should toggle the bulk action menu, and all the actions should work as expected.

<img width="1566" alt="Screenshot 2023-01-17 at 12 25 11 PM" src="https://user-images.githubusercontent.com/10586875/212830451-5bd9f673-f189-4e56-bedb-6c71a04a4618.png">

<img width="1585" alt="Screenshot 2023-01-17 at 12 25 03 PM" src="https://user-images.githubusercontent.com/10586875/212830490-33ad2ec2-a0cf-4072-92b0-58a4770c22d0.png">

https://user-images.githubusercontent.com/10586875/212829910-06694891-d6f7-4647-9d5b-9f038c27dddd.mov

4. Switch the small screen view using the dev tool and verify step 3 works as expected.

<img width="368" alt="Screenshot 2023-01-17 at 12 26 42 PM" src="https://user-images.githubusercontent.com/10586875/212830530-0b3bd9f5-3381-4ca7-b68e-1a02e99805ea.png">

<img width="359" alt="Screenshot 2023-01-17 at 12 26 54 PM" src="https://user-images.githubusercontent.com/10586875/212830539-aed529e0-db9c-4c9e-afd1-2d7f1d1260e2.png">

<img width="368" alt="Screenshot 2023-01-17 at 12 26 58 PM" src="https://user-images.githubusercontent.com/10586875/212830542-7f82bd5d-df11-4cbd-9b27-43105ca47fb6.png">

https://user-images.githubusercontent.com/10586875/212830229-890e925a-cda7-42cf-83ff-83ca39f86e8a.mov

5. Set the feature flag `jetpack/partner-portal-downtime-monitoring-updates` to false and verify it works as in production.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 1203448324265423-as-1203507036638751

DustinHartzler committed:
Added Edit All buttons - still needs formatting

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -12,22 +16,69 @@ interface Props {
+	 * this logic take the existing state and inverts it,
+	 * I've used the letter b because it's a Boolean
+	 */
+	const toggleBulkEditMode = () => setIsBulkManagementActive( ( b ) => ! b );
```
How about we can keep it `isActive` instead of `b`?

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -12,22 +16,69 @@ interface Props {
+					</tr>
+				) : (
+					<tr>
+						<ButtonGroup className="edit-all__bulk-actions-pause-monitor">
```
We can wrap everything inside the `tr` in a `th` tag with `colSpan={ 100 }` and `text-align: right`

```
<tr>
    <th colSpan={100}>
       <ButtonGroup>
       ..........
   </th>
</tr>
```

yashwin replied:
We could take some inspiration from https://github.com/Automattic/wp-calypso/blob/add/jetpack-dashboard-downtime-monitoring-bulk-action-menu/client/my-sites/plugins/plugin-list-header/index.jsx. This is how it is being done in the Plugin's page where we switch to a dropdown in smaller screens.

DustinHartzler replied:
Exactly what I used as a model. Thanks!

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -12,22 +16,69 @@ interface Props {
+								{ translate( 'Pause Monitor', { context: 'button label' } ) }
+							</Button>
+						</ButtonGroup>
+						<ButtonGroup className="edit-all__bulk-actions-resume-monitor">
```
I don't think we need class name for `ButtonGroup`s

DustinHartzler replied:
I had to leave the classes, so I could style them appropriately.

yashwin replied:
I don't think it is needed for all the button groups.

DustinHartzler replied:
You are correct. I removed them as they were unnecessary.

DustinHartzler committed:
Added Mobile Bulk Menu

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -1,33 +1,86 @@
+							</Button>
 						</th>
-					) ) }
-					<th></th>
```
We will need this to keep an empty space.

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -1,33 +1,88 @@
 }
 
 export default function SiteTable( { isLoading, columns, items }: Props ) {
+	//const initialText = 'Edit All';
```
Can we remove this comment?

DustinHartzler replied:
Yes. Done.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/style.scss:
```diff
@@ -1,3 +1,46 @@
+
+.edit-all__bulk-actions {
+	margin: 10px;
+	padding-right: 5px;
```
Could we use RTL-aware style attributes everywhere on this file? 

RTL equivalent properties

padding-right: padding-inline-end
padding-left: padding-inline-start
padding-top: padding-block-start
padding-bottom: padding-block-end

The same goes for margin.

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/style.scss:
```diff
@@ -1,3 +1,46 @@
+}
+
+.button-group {
+	margin-top: 6px;
```
margin-block-start*

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/style.scss:
```diff
@@ -1,3 +1,46 @@
+}
+
+th.edit-all-buttons {
+	text-align: right;
```
Could we use RTL-aware style attributes everywhere on this file? 

text-align: start/end

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-table/index.tsx:
```diff
@@ -1,33 +1,88 @@
+						) ) }
+						{ /* // This is needed for the feature flag */ }
+						<th></th>
+						{ ! isEnabled( 'jetpack/partner-portal-downtime-monitoring-updates' ) && (
```
Should this be `isEnabled` instead of `!isEnabled`?

yashwin reviewed (changes requested):
I reviewed this PR and found a couple of issues.

1) The buttons' styles should match how it looks on the Plugins page.

Dashboard
<img width="583" alt="Screenshot 2023-01-10 at 4 01 20 PM" src="https://user-images.githubusercontent.com/10586875/211527747-acd90ddb-bd57-45c3-a40e-e693c4d7a05d.png">

Plugins
<img width="583" alt="Screenshot 2023-01-10 at 3 58 11 PM" src="https://user-images.githubusercontent.com/10586875/211527006-785cb3e4-6d68-409c-bb58-3b55789ad72d.png">

Including hover/active styles

Dashboard
<img width="133" alt="Screenshot 2023-01-10 at 3 56 51 PM" src="https://user-images.githubusercontent.com/10586875/211526707-1bccd9ba-8499-44dc-8fb7-9da4d5aff099.png">

Plugins
<img width="80" alt="Screenshot 2023-01-10 at 4 00 43 PM" src="https://user-images.githubusercontent.com/10586875/211527510-3c859054-87b1-49d1-812f-
4112297d587e.png">

2) I tried enabling `Edit All`, and there is empty space below.

<img width="409" alt="Screenshot 2023-01-10 at 4 03 35 PM" src="https://user-images.githubusercontent.com/10586875/211528302-39aefa3c-4231-4f3c-bc70-5b4bc2eec94e.png">

3) I feel that the bulk actions menu should adjust based on the content rather than switching to dropdown only on mobile view like it is done on the Plugins page - https://github.com/Automattic/wp-calypso/blob/add/jetpack-dashboard-downtime-monitoring-bulk-action-menu/client/my-sites/plugins/plugin-list-header/index.jsx. It breaks in other languages

<img width="610" alt="Screenshot 2023-01-10 at 4 05 44 PM" src="https://user-images.githubusercontent.com/10586875/211529176-3200469c-0a1f-4332-87db-287d12674b63.png">

DustinHartzler committed:
Updated CSS to match plugins page

DustinHartzler commented:
I feel like the plugin section isn't working either when the buttons get to be wider than the column width:
![Screenshot 2023-01-10 at 13 27 59](https://user-images.githubusercontent.com/2293077/211914407-1ce223b9-d230-4c19-b4b7-d6ec816fa8f9.png)

Any ideas on how to solve? Or should this be an enhancement to address both areas?

yashwin commented:
> I feel like the plugin section isn't working either when the buttons get to be wider than the column width

I think you tried changing the HTML content, which will not work since this is done at the JS level. 

This is how it looks on a large screen on the Plugins page.

<img width="673" alt="Screenshot 2023-01-12 at 11 10 17 AM" src="https://user-images.githubusercontent.com/10586875/211986697-2a7759f8-618a-496b-8551-b1b589dc40cf.png">

> Any ideas on how to solve? Or should this be an enhancement to address both areas?

We can take some inspiration on how it is done here - https://github.com/Automattic/wp-calypso/blob/add/jetpack-dashboard-downtime-monitoring-bulk-action-menu/client/my-sites/plugins/plugin-list-header/index.jsx

We have to do something similar to this; we have to make some changes since the below is done in a class-based component, and we are using a function-based component.

<img width="711" alt="Screenshot 2023-01-12 at 11 13 05 AM" src="https://user-images.githubusercontent.com/10586875/211987117-ede54a69-b458-4f8c-800b-948fc0155310.png">

yashwin reviewed (commented):
I did another round of review, looks much better now.

I could still see some issues

1) The action names are not correct.
2) The hover style is not fixed.

<img width="661" alt="Screenshot 2023-01-12 at 11 31 10 AM" src="https://user-images.githubusercontent.com/10586875/211990125-c0219e92-7c8c-4704-9dab-4fb654fad8c7.png">

3) I could still see the empty space.

<img width="534" alt="Screenshot 2023-01-12 at 11 35 06 AM" src="https://user-images.githubusercontent.com/10586875/211990291-b0515d0a-acc2-4522-b1dc-8963244b4aa2.png">

4) I see this alignment issue on a smaller screen(961px)

<img width="691" alt="Screenshot 2023-01-12 at 11 38 37 AM" src="https://user-images.githubusercontent.com/10586875/211990791-9732cb3f-beb5-4030-a1a2-e0baeb0ab664.png">

DustinHartzler commented:
> 1. The action names are not correct.

This was a typo from checking out the layout issues. I've fixed this.
 
> 2. The hover style is not fixed.

I confirmed with Kristian and he is seeing the proper hover colors

> 3.  I could still see the empty space.

The empty column is there because of the `<th></th>` that's added to keep the live version looking okay. This needs to be removed to do the testing.

> 4. I see this alignment issue on a smaller screen(961px)

It seems like the header of the table was quite crowded between 960px - 1040px, so I updated the mobile breakpoint to be changed at 1040px. Let me know if there's a better solution.


> We can take some inspiration on how it is done here - https://github.com/Automattic/wp-calypso/blob/add/jetpack-dashboard-downtime-monitoring-bulk-action-menu/client/my-sites/plugins/plugin-list-header/index.jsx

> We have to do something similar to this; we have to make some changes since the below is done in a class-based component, and we are using a function-based component.

I wasn't able to get to this today.

DustinHartzler committed:
Small fixes

yashwin commented:
> I confirmed with Kristian and he is seeing the proper hover colors

I wanted to mention `:active` and typed it `hover`. Apologies for the same.

yashwin reviewed (commented):
I had another look at this PR. Looks much better now

A couple of things I noticed

1) The actions(`...`) should be aligned to the very right. Check https://cloud.jetpack.com/ to see how it looks.

<img width="305" alt="Screenshot 2023-01-13 at 11 27 57 AM" src="https://user-images.githubusercontent.com/10586875/212248152-02034e3c-674a-47a5-bf9a-5493f5d96fe5.png">

2) The actions and issue count is not visible from 961-1024px

<img width="372" alt="Screenshot 2023-01-13 at 11 26 02 AM" src="https://user-images.githubusercontent.com/10586875/212248425-8b47f3b5-942d-491a-b166-d2f9f293e95e.png">

yashwin commented:
> It seems like the header of the table was quite crowded between 960px - 1040px, so I updated the mobile breakpoint to be changed at 1040px. Let me know if there's a better solution.

We can do it. But, let's also check with the design team. @madebynoam, any thoughts here? We will show the mobile view until 1024px instead of 960px. I think we could do it until 1080px since that's what we are doing on the Plugins page. We can keep it consistent.

madebynoam commented:
> I think we could do it until 1080px since that's what we are doing on the Plugins page. We can keep it consistent.

Hey, @yashwin, that sounds good to me! The more we can be consistent with the plugins page, the better.

DustinHartzler committed:
Updated mobile breakpoints

DustinHartzler commented:
> The actions(...) should be aligned to the very right. Check https://cloud.jetpack.com/ to see how it looks.

>  The actions and issue count is not visible from 961-1024px

Both of these have been fixed.

Have also set the mobile vs desktop break point at 1080px.

Moving the task to change the buttons to a drop down when the translated text is too long to another task: 1203448324265423-as-1203706311193972/f

yashwin committed:
UI fixes for large & small screen views

yashwin committed:
UI enhancements & fixes

vitozev reviewed (approved):
LGTM! I confirm the changes look great on both desktop and mobile, and no regression was found.
